doc/developer/design/20230512_mir_cost_model.md

```diff
@@ -102,6 +104,25 @@ This hasn't been implemented yet.
 # Rollout
 [rollout]: #rollout
 
+## Implementation plan
+[testing-and-observability]: #implementation-plan
+
+The current plan is to work towards a minimally viable cost model, with the following priority ordering memory > hydration > latency > scalability.
+
+- Start with a source cardinality analysis that maps (symbolic) input cardinalities to (symbolic) output cardinalities.
+
+  + Cardinality analysis can work on MIR or LIR, since cardinality ought to be constant under optimization.
+
+  + Selectivity factors will need to be just made up for now.
```

Comment:
@danhhz / @dseisun-materialize / @bkirwi If the system catalog can expose an aggregated (even if eventually consistent) view over aggregated block-level statistics, we might be able to start using those to infer basic statistics.

Reply:
Can definitely imagine exposing something from the storage controller. We should chat sometime about what a good interface for that would look like. It's likely we could provide not only current stats, but also some measure of how rapidly those stats are changing...

Reply:
Also conceivable that we could provide stats for the "post-filter-pushdown" version of the dataset... eg. if we apply a filter at the source that allows us to skip fetching 90% of parts, we could probably give you the stats for just the remaining 10%. This may help address some issues around scaling factors that @vmarcos mentioned earlier. (Since at least in some cases we can apply a filter and still keep a good upper bound.) But that's a little speculative and I'm not sure it's general enough to be useful... there are many filters that might be highly selective and yet not susceptible to filter pushdown.

Reply:
yeah, storage is on board with maintaining the necessary stats for cost estimation. i'm not sure i'm confidant that i understand precisely the details of what y'all need yet, though

@bkirwi the idea being that we could run the pushdown code and hand back stats for the parts that didn't get skipped? that's a very cool idea. can't hurt and might help a lot

Reply:
Yep that's right!